File: domains/ext-authz/internal/tracing/tracing.go
// Package tracing provides OpenTelemetry distributed tracing configuration.
//
// Architecture (CNCF Best Practices):
//
//	App (OTel SDK) → OTLP/gRPC (4317) → Jaeger Collector → Elasticsearch
//
// References:
//   - Google Dapper: Low overhead, application-level transparency
//   - Netflix Edgar: Request interceptor pattern, async collection
//   - Uber Jaeger: OpenTelemetry native, adaptive sampling
package tracing

import (
	"context"
	"os"
	"strconv"
	"time"

	"go.opentelemetry.io/contrib/propagators/b3"
	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/codes"
	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
	"go.opentelemetry.io/otel/propagation"
	"go.opentelemetry.io/otel/sdk/resource"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.26.0"
	"go.opentelemetry.io/otel/trace"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"

	"github.com/eco2-team/backend/domains/ext-authz/internal/constants"
)

// Default configuration
const (
	defaultEndpoint     = "jaeger-collector-clusterip.istio-system.svc.cluster.local:4317"
	defaultSamplingRate = 1.0
	defaultTimeout      = 5 * time.Second
)

// Environment variable names
const (
	envEndpoint     = "OTEL_EXPORTER_OTLP_ENDPOINT"
	envSamplingRate = "OTEL_SAMPLING_RATE"
	envEnabled      = "OTEL_ENABLED"
)

// Config holds tracing configuration.
type Config struct {
	ServiceName    string
	ServiceVersion string
	Environment    string
	Endpoint       string
	SamplingRate   float64
	Enabled        bool
}

// DefaultConfig returns default tracing configuration from environment.
func DefaultConfig() *Config {
	enabled := true
	if v := os.Getenv(envEnabled); v != "" {
		enabled = v == "true"
	}

	samplingRate := defaultSamplingRate
	if v := os.Getenv(envSamplingRate); v != "" {
		if f, err := strconv.ParseFloat(v, 64); err == nil {
			samplingRate = f
		}
	}

	endpoint := os.Getenv(envEndpoint)
	if endpoint == "" {
		endpoint = defaultEndpoint
	}

	return &Config{
		ServiceName:    constants.ServiceName,
		ServiceVersion: constants.ServiceVersion,
		Environment:    os.Getenv(constants.EnvEnvironment),
		Endpoint:       endpoint,
		SamplingRate:   samplingRate,
		Enabled:        enabled,
	}
}

// TracerProvider wraps the OpenTelemetry TracerProvider.
type TracerProvider struct {
	provider *sdktrace.TracerProvider
}

// Init initializes OpenTelemetry tracing.
func Init(ctx context.Context, cfg *Config) (*TracerProvider, error) {
	if cfg == nil {
		cfg = DefaultConfig()
	}

	if !cfg.Enabled {
		return nil, nil
	}

	// Create OTLP gRPC exporter
	conn, err := grpc.NewClient(
		cfg.Endpoint,
		grpc.WithTransportCredentials(insecure.NewCredentials()),
	)
	if err != nil {
		return nil, err
	}

	exporter, err := otlptracegrpc.New(ctx, otlptracegrpc.WithGRPCConn(conn))
	if err != nil {
		return nil, err
	}

	// Create resource with service metadata
	// Use resource.New instead of resource.Merge to avoid schema URL conflicts
	res, err := resource.New(ctx,
		resource.WithAttributes(
			semconv.ServiceName(cfg.ServiceName),
			semconv.ServiceVersion(cfg.ServiceVersion),
			attribute.String("deployment.environment", cfg.Environment),
			attribute.String("telemetry.sdk.name", "opentelemetry"),
			attribute.String("telemetry.sdk.language", "go"),
		),
		resource.WithHost(),
		resource.WithProcess(),
	)
	if err != nil {
		return nil, err
	}

	// Create sampler (ratio-based for production)
	sampler := sdktrace.TraceIDRatioBased(cfg.SamplingRate)

	// Create TracerProvider
	tp := sdktrace.NewTracerProvider(
		sdktrace.WithResource(res),
		sdktrace.WithSampler(sampler),
		sdktrace.WithBatcher(exporter,
			sdktrace.WithMaxQueueSize(2048),
			sdktrace.WithMaxExportBatchSize(512),
			sdktrace.WithBatchTimeout(time.Second),
		),
	)

	// Set global TracerProvider and propagator
	// B3 propagator for Istio/Envoy compatibility
	otel.SetTracerProvider(tp)
	otel.SetTextMapPropagator(propagation.NewCompositeTextMapPropagator(
		propagation.TraceContext{},
		propagation.Baggage{},
		b3.New(b3.WithInjectEncoding(b3.B3MultipleHeader)),
	))

	return &TracerProvider{provider: tp}, nil
}

// Shutdown gracefully shuts down the tracer provider.
func (tp *TracerProvider) Shutdown(ctx context.Context) error {
	if tp == nil || tp.provider == nil {
		return nil
	}

	ctx, cancel := context.WithTimeout(ctx, defaultTimeout)
	defer cancel()

	return tp.provider.Shutdown(ctx)
}

// Tracer returns a tracer for the given name.
func Tracer(name string) trace.Tracer {
	return otel.Tracer(name)
}

// StartSpan starts a new span with the given name and attributes.
func StartSpan(ctx context.Context, name string, attrs ...attribute.KeyValue) (context.Context, trace.Span) {
	return Tracer("ext-authz").Start(ctx, name, trace.WithAttributes(attrs...))
}

// SpanFromContext returns the current span from context.
func SpanFromContext(ctx context.Context) trace.Span {
	return trace.SpanFromContext(ctx)
}

// AddEvent adds an event to the current span.
func AddEvent(ctx context.Context, name string, attrs ...attribute.KeyValue) {
	span := trace.SpanFromContext(ctx)
	span.AddEvent(name, trace.WithAttributes(attrs...))
}

// SetError marks the span as an error.
func SetError(ctx context.Context, err error, description string) {
	span := trace.SpanFromContext(ctx)
	span.RecordError(err)
	span.SetStatus(codes.Error, description)
}

// RecordError records an error on the current span.
func RecordError(ctx context.Context, err error) {
	span := trace.SpanFromContext(ctx)
	span.RecordError(err)
}

File: apps/character/infrastructure/observability/tracing.py
"""OpenTelemetry Tracing - Character Service.

domains 의존성 제거 - 내부 모듈.
"""

import logging
import os

logger = logging.getLogger(__name__)

OTEL_EXPORTER_ENDPOINT = os.getenv(
    "OTEL_EXPORTER_OTLP_ENDPOINT",
    "jaeger-collector.istio-system.svc.cluster.local:4317",
)
OTEL_SAMPLING_RATE = float(os.getenv("OTEL_SAMPLING_RATE", "1.0"))
OTEL_ENABLED = os.getenv("OTEL_ENABLED", "true").lower() == "true"


def setup_tracing(service_name: str) -> bool:
    """OpenTelemetry 트레이싱 설정.

    Args:
        service_name: 서비스 이름

    Returns:
        설정 성공 여부
    """
    if not OTEL_ENABLED:
        logger.info("OpenTelemetry tracing disabled (OTEL_ENABLED=false)")
        return False

    try:
        from opentelemetry import trace
        from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import (
            OTLPSpanExporter,
        )
        from opentelemetry.sdk.resources import Resource
        from opentelemetry.sdk.trace import TracerProvider
        from opentelemetry.sdk.trace.export import BatchSpanProcessor
        from opentelemetry.sdk.trace.sampling import TraceIdRatioBased

        resource = Resource.create(
            {
                "service.name": service_name,
                "service.version": os.getenv("SERVICE_VERSION", "1.0.0"),
                "deployment.environment": os.getenv("ENVIRONMENT", "dev"),
            }
        )

        sampler = TraceIdRatioBased(OTEL_SAMPLING_RATE)
        provider = TracerProvider(resource=resource, sampler=sampler)

        exporter = OTLPSpanExporter(
            endpoint=OTEL_EXPORTER_ENDPOINT,
            insecure=True,
        )

        provider.add_span_processor(
            BatchSpanProcessor(
                exporter,
                max_queue_size=2048,
                max_export_batch_size=512,
                schedule_delay_millis=1000,
            )
        )

        trace.set_tracer_provider(provider)

        logger.info(
            "OpenTelemetry tracing configured",
            extra={
                "service": service_name,
                "endpoint": OTEL_EXPORTER_ENDPOINT,
                "sampling_rate": OTEL_SAMPLING_RATE,
            },
        )
        return True

    except ImportError as e:
        logger.warning(f"OpenTelemetry not available: {e}")
        return False
    except Exception as e:
        logger.error(f"Failed to configure tracing: {e}")
        return False


def instrument_fastapi(app) -> None:
    """FastAPI 자동 계측."""
    if not OTEL_ENABLED:
        return

    try:
        from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor

        FastAPIInstrumentor.instrument_app(
            app,
            excluded_urls="health,ready,metrics",
        )
        logger.info("FastAPI instrumentation enabled")

    except ImportError:
        logger.warning("FastAPIInstrumentor not available")
    except Exception as e:
        logger.error(f"Failed to instrument FastAPI: {e}")


def instrument_httpx() -> None:
    """HTTPX 자동 계측."""
    if not OTEL_ENABLED:
        return

    try:
        from opentelemetry.instrumentation.httpx import HTTPXClientInstrumentor

        HTTPXClientInstrumentor().instrument()
        logger.info("HTTPX instrumentation enabled")

    except ImportError:
        logger.warning("HTTPXClientInstrumentor not available")
    except Exception as e:
        logger.error(f"Failed to instrument HTTPX: {e}")


def instrument_redis() -> None:
    """Redis 자동 계측 (캐시 추적)."""
    if not OTEL_ENABLED:
        return

    try:
        from opentelemetry.instrumentation.redis import RedisInstrumentor

        RedisInstrumentor().instrument()
        logger.info("Redis instrumentation enabled")

    except ImportError:
        logger.warning("RedisInstrumentor not available")
    except Exception as e:
        logger.error(f"Failed to instrument Redis: {e}")


def shutdown_tracing() -> None:
    """트레이싱 종료."""
    # TracerProvider는 global이므로 shutdown은 필요하지 않음
    # 향후 _tracer_provider를 저장하여 shutdown 구현 가능
    pass
